### PR TITLE
XFCC header support

### PR DESCRIPTION
**Pull request by gautierdelorme**
This is required to let apps use details from client certificates (e.g. Subject, SAN...). Since the certificate (or the certificate chain) could exceed the web server header size limit, we give the ability to select what specific part of the certificate to expose in the `x-forwarded-client-cert` header.

Fixes https://github.com/projectcontour/contour/issues/2885 (I've read through this issue and since that feature is only usable when `ClientValidation` is used (since that's required for the server to request certs) I think having `ForwardClientCertificate` under `ClientValidation` makes more sense)

Signed-off-by: Gautier Delorme <gautier.delorme@gmail.com>

### CODE REVIEW

**Comment by gautierdelorme** [comment]
Do you think this change could make it to the 1.23 release?

**Comment by skriss** [comment]
> Do you think this change could make it to the 1.23 release?

Unfortunately it's too late to get these into 1.23, but we'll pick up with reviews shortly.

**Review by tsaarni — CHANGES_REQUESTED** [review]
Looking very good :+1: Just one tiny remark from me.

File: internal/dag/dag.go [hunk 1]
```diff
@@ -523,6 +523,25 @@ type HeaderValue struct {
 	Value string
 }
 
+// ClientCertificateDetails defines which parts of the client certificate will be forwarded.
+type ClientCertificateDetails struct {
+	// Subject of the client cert.
+	// +optional
```
[the remark this answers is not in the document]
Comment (gautierdelorme) [reply internal/dag/dag.go]: Totally 👍 Fixed in https://github.com/projectcontour/contour/pull/4797/commits/b7e8dabc5939ecb37b162751baf3521421cc0417.

**Review by tsaarni — CHANGES_REQUESTED** [review]
Needs merging & regenerating after the other PR was merged, but otherwise ready to go from my perspective.

**Review by tsaarni — APPROVED** [review]
Thank you @gautierdelorme :+1: 
Leaving for others to review as well.

**Review by skriss — APPROVED** [review]
LGTM, thanks @gautierdelorme! Will leave open for a bit in case @sunjayBhatia wants to take a look.

**Review by sunjayBhatia — APPROVED** [review]
looks great!

one thing that could be useful to help troubleshoot header size issues would be to enable upstream cluster req/resp statistics: https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#request-response-size-statistics

I can add a new issue for that so we can consider it (if we dont have one already)

**Pull request merged**